extensions/amp-analytics/0.1/session-manager.js

```diff
@@ -0,0 +1,252 @@
+      this.sessions_[type].lastAccessTimestamp = Date.now();
+      return Promise.resolve(this.sessions_[type]);
+    }
+
+    return this.getSessionFromStorage_(type);
+  }
+
+  /**
+   * Get our session if it exists or creates it. Sets the session
+   * in localStorage to update the last access time.
+   * @param {string} type
+   * @return {!Promise<SessionInfoDef>}
+   */
+  getSessionFromStorage_(type) {
+    return this.storagePromise_
+      .then((storage) => {
+        const storageKey = getStorageKey(type);
+        // Gets the session if it has not expired
+        return storage.get(storageKey, SESSION_MAX_AGE_MILLIS);
+      })
+      .then((session) => {
+        // If no valid session in storage, create a new one
+        return !session
+          ? constructSessionInfo(generateSessionId(), Date.now())
+          : constructSessionFromStoredValue(session);
+      })
+      .then((session) => {
+        // Avoid multiple session creation race
+        if (
+          type in this.sessions_ &&
+          !this.isSessionExpired_(this.sessions_[type])
+        ) {
+          return this.sessions_[type];
+        }
+        this.setSession_(type, session);
+        session.lastAccessTimestamp = Date.now();
+        this.sessions_[type] = session;
+        return this.sessions_[type];
+      });
+  }
+
+  /**
+   * Set the session in localStorage, updating
+   * its last access time if it did not exist before.
+   * @param {string} type
+   * @param {SessionInfoDef} session
+   * @return {!Promise}
+   */
+  setSession_(type, session) {
+    return this.storagePromise_.then((storage) => {
+      const storageKey = getStorageKey(type);
+      const value = composeStorageSessionValue(session);
+      storage.setNonBoolean(storageKey, value);
+    });
+  }
+
+  /**
+   * Checks if a session has expired
+   * @param {SessionInfoDef} session
+   * @return {!Promise}
+   */
+  isSessionExpired_(session) {
+    return session.lastAccessTimestamp + SESSION_MAX_AGE_MILLIS < Date.now();
+  }
+}
+
+/**
+ * Return a pseudorandom low entropy value for session id.
+ * @return {number}
+ */
+function generateSessionId() {
+  return Math.floor(10000 * Math.random());
+}
+
+/**
+ * @param {string} type
+ * @return {string}
+ */
+function getStorageKey(type) {
+  return SESSION_STORAGE_KEY + type;
+}
+
+/**
+ * @param {SessionInfoDef|string} storedSession
+ * @return {SessionInfoDef|undefined}
+ */
+function constructSessionFromStoredValue(storedSession) {
+  if (!isObject(storedSession)) {
+    dev().error(TAG, 'Invalid stored session value');
+    return;
+  }
+
+  return constructSessionInfo(
+    storedSession[SESSION_STORAGE_KEYS.SESSION_ID],
+    storedSession[SESSION_STORAGE_KEYS.CREATION_TIMESTAMP],
+    Date.now()
+  );
+}
+
+/**
+ * @param {!SessionInfoDef} session
+ * @return {!Object}
+ */
+export function composeStorageSessionValue(session) {
+  const obj = map();
+  obj[SESSION_STORAGE_KEYS.SESSION_ID] = session.sessionId;
```

Comment:
Yes they could be. In the past, we used a mapping so that we could save bits in localStorage and to potentially obfuscate the meaning of our kv pairs. While we no longer have a size restriction on localStroage in the viewer, I think it's still not a bad idea to keep this mapping, however, it will add some size to our binary (which gets served often).

What do you think we should do?

Reply:
Could you use the same enum values like `SESSION_STORAGE_KEYS.SESSION_ID` consistently? That way you can use whatever size keys you want, and keep the object unmodified.

Reply:
Done